cosmos.go

```diff
@@ -295,40 +303,49 @@ func (c *cosmosImpl) retryLoop(executeRequest retryFun) (responses []interfaces.
 		}
 
 		if retryInformation.retryAfter > 0 {
-			c.logger.Info().Msgf("retry %d of query after %v because of header status code %d", tryCount+1, retryInformation.retryAfter, retryInformation.responseStatusCode)
+			logger.Info().Msgf("retry %d of query after %v because of header status code %d", tryCount+1, retryInformation.retryAfter, retryInformation.responseStatusCode)
 
 			if waitDone := waitForRetry(retryInformation.retryAfter, timeoutReachedChan); !waitDone {
-				break
+				// timeout occurred
+				metrics.requestRetryTimeoutsTotal.Inc()
+				return nil, fmt.Errorf("timed out while waiting to do a retry after %s (timeout=%s)", retryInformation.retryAfter, retryTimeout)
 			}
 		}
 
 		// Timeout check in case no waiting is required
 		select {
 		case <-timeoutReachedChan:
 			// we stop here and return what we got so far
-			return responses, err
+			metrics.requestRetryTimeoutsTotal.Inc()
+			return nil, fmt.Errorf("timed out while doing a retry (timeout=%s)", retryTimeout)
 		default:
+			continue
 			// continue with next retry
 		}
 	}
 
 	return responses, err
 }
 
-func (c *cosmosImpl) handleTimeout(done <-chan bool) (timedOutChan <-chan bool) {
-
-	timeoutReachedChan := make(chan bool)
+func handleTimeout(done <-chan struct{}, retryTimeout time.Duration, logger zerolog.Logger) (timedOutChan <-chan struct{}) {
+	timeoutReachedChan := make(chan struct{})
```

Comment:
This changes the results of this producer, as the channel closed in both cases (timedOut, done) and thus the caller can't determine which situation occurred.

Reply:
will check to switch back to bool chan